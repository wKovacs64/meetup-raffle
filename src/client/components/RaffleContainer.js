import React from 'react';
import axios from 'axios';
import get from 'lodash/get';
import { RingLoader } from 'react-spinners';
import { Formik } from 'formik';
import { restore, preserve } from '../persistence';
import ErrorMessage from './ErrorMessage';
import RaffleForm from './RaffleForm';
import ResetButtons from './ResetButtons';
import Results from './Results';

const initialFormValues = {
  meetup: '',
  count: '1',
};

const initialResults = {
  error: '',
  winners: [],
};

const RaffleContainer = () => {
  const [count, setCount] = React.useState(initialFormValues.count);
  const [meetup, setMeetup] = React.useState(initialFormValues.meetup);
  const [error, setError] = React.useState(initialResults.error);
  const [winners, setWinners] = React.useState(initialResults.winners);

  const restoreSettings = () => {
    const storedMeetup = restore('meetup');
    if (storedMeetup) {
      setMeetup(storedMeetup);
    }

    const storedCount = restore('count');
    if (storedCount) {
      setCount(storedCount);
    }
  };

  React.useEffect(() => {
    restoreSettings();
  }, []);

  const reset = () => {
    restoreSettings();
    setError(initialResults.error);
    setWinners(initialResults.winners);
  };

  const handleApiError = err => {
    setError(get(err, 'response.data.error.message', err.message));
  };

  const handleFormikSubmit = async (
    { meetup: meetupValue, count: countValue },
    { setSubmitting },
  ) => {
    setSubmitting(true);
    preserve({ meetup: meetupValue, count: countValue });
    try {
      const response = await axios.get('/.netlify/functions/draw', {
        params: { meetup: meetupValue, count: countValue },
      });
      const winnersData = get(response, 'data.winners');
      if (winnersData) {
        setWinners(winnersData);
      } else {
        throw new Error('Malformed response received.');
      }
    } catch (err) {
      handleApiError(err);
    }
    setSubmitting(false);
  };

  return (
    <main className="flex flex-column flex-grow-1 flex-shrink-0 ph3 w-100 mw6-m mw7-l self-center-ns">
      <Formik
        enableReinitialize
        initialValues={{ meetup, count }}
        onSubmit={handleFormikSubmit}
      >
        {({ handleSubmit, isSubmitting, setFieldValue }) => {
          if (isSubmitting) {
            return (
              <div className="flex flex-grow-1 flex-shrink-0 justify-center items-center items-start-ns mt3 mt4-ns">
                <div className="h4 w4">
                  <RingLoader size={128} color="#00449e" />
                </div>
              </div>
            );
          }
          if (error) {
            return (
              <div className="mt3 mt4-ns">
                <ErrorMessage problemText={error} data-testid="error-message" />
                <ResetButtons onReset={reset} onSubmit={handleSubmit} />
              </div>
            );
          }
          if (winners.length) {
            return (
              <div className="mt3 mt4-ns">
                <Results winners={winners} data-testid="results" />
                <ResetButtons onReset={reset} onSubmit={handleSubmit} />
              </div>
            );
          }
          return (
            <RaffleForm
              defaultCount={parseInt(count, 10)}
              setFieldValue={setFieldValue}
            />
          );
        }}
      </Formik>
    </main>
  );
};

RaffleContainer.displayName = 'RaffleContainer';

export default RaffleContainer;
