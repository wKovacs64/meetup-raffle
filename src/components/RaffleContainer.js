import React, { Component, Fragment } from 'react';
import axios from 'axios';
import get from 'lodash.get';
import { Formik } from 'formik';
import { ErrorMessage, Loading, RaffleForm, ResetButtons, Results } from '.';

export class RaffleContainer extends Component {
  // eslint-disable-next-line react/sort-comp
  initialResults = {
    error: '',
    winners: [],
  };

  initialFormValues = {
    meetup: '',
    count: 1,
    specificEventId: '',
    meetupApiKey: '',
  };

  state = {
    ...this.initialResults,
    ...this.initialFormValues,
  };

  componentDidMount() {
    const meetup = this.restore('meetup');
    const count = this.restore('count');
    const meetupApiKey = this.restore('meetupApiKey');

    this.setState({
      ...(meetup && { meetup }),
      ...(count && { count }),
      ...(meetupApiKey && { meetupApiKey }),
    });
  }

  resetResults = () => {
    this.setState(this.initialResults);
  };

  restore = key => {
    if (global.window.localStorage) {
      return global.window.localStorage.getItem(key);
    }
    return undefined;
  };

  preserve = data => {
    if (global.window.localStorage) {
      Object.entries(data).forEach(([key, value]) => {
        global.window.localStorage.setItem(key, value);
      });
    }
  };

  handleApiError = err => {
    const error = get(err, 'response.data.error.message', err.message);
    this.setState({ error });
  };

  handleFormikSubmit = async (
    { meetup, count, specificEventId, meetupApiKey },
    { setSubmitting },
  ) => {
    setSubmitting(true);
    this.preserve({
      meetup,
      count,
      meetupApiKey,
    });
    try {
      const response = await axios.get(
        'https://wkovacs64.lib.id/meetup-raffle/',
        {
          params: {
            meetup,
            count,
            specificEventId,
            meetupApiKey,
          },
        },
      );
      const winners = get(response, 'data.winners');
      if (winners) {
        this.setState({ winners });
      } else {
        throw new Error('Malformed response received.');
      }
    } catch (err) {
      this.handleApiError(err);
    }
    setSubmitting(false);
  };

  renderFormik = ({ handleSubmit, isSubmitting }) => {
    if (isSubmitting) {
      return (
        <div className="tc">
          <Loading className="h5 w5 dark-blue" />
        </div>
      );
    }
    if (this.state.error) {
      return (
        <Fragment>
          <ErrorMessage problemText={this.state.error} />
          <ResetButtons onReset={this.resetResults} onSubmit={handleSubmit} />
        </Fragment>
      );
    }
    if (this.state.winners.length) {
      return (
        <Fragment>
          <Results winners={this.state.winners} />
          <ResetButtons onReset={this.resetResults} onSubmit={handleSubmit} />
        </Fragment>
      );
    }
    return <RaffleForm />;
  };

  render() {
    return (
      <main className="ph3 pv3 pv4-ns mw6-m mw7-l center-ns">
        <Formik
          enableReinitialize
          initialValues={{
            meetup: this.state.meetup,
            count: this.state.count,
            specificEventId: this.initialFormValues.specificEventId,
            meetupApiKey: this.state.meetupApiKey,
          }}
          onSubmit={this.handleFormikSubmit}
          render={this.renderFormik}
        />
      </main>
    );
  }
}

export default RaffleContainer;
