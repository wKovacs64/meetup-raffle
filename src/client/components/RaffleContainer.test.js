import React from 'react';
import {
  render,
  fireEvent,
  waitForElement,
  wait,
} from '@testing-library/react';
import mockAxios from 'axios';
import RaffleContainer from './RaffleContainer';

const mockWinners = Array.from(Array(2), (_, idx) => ({
  name: `Pickle Rick ${idx}`,
  photoURL: `https://i.imgur.com/3VhMoBD.png?i=${idx}`,
  profileURL: `https://en.wikipedia.org/wiki/Pickle_Rick?i=${idx}`,
}));

const params = {
  meetup: 'foo',
  count: '2',
};

describe('RaffleContainer', () => {
  const { localStorage } = global.window;

  const fillOutForm = ({ getByLabelText }) => {
    // find elements
    const meetupInput = getByLabelText(/Meetup name/);
    const countInput = getByLabelText('Number of winners:');

    // fill out form
    fireEvent.change(meetupInput, { target: { value: params.meetup } });
    fireEvent.change(countInput, { target: { value: params.count } });
  };

  const submitForm = async ({ getByText }) => {
    const drawButton = getByText('Draw');

    // submit form
    fireEvent.click(drawButton);
  };

  beforeEach(() => {
    jest.clearAllMocks();
    global.window.localStorage.clear();
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it('renders', () => {
    const { container } = render(<RaffleContainer />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('restores data from localStorage (if available)', () => {
    const firstRender = render(<RaffleContainer />);
    expect(
      parseInt(firstRender.getByLabelText(/Number of winners/i).value, 10),
    ).not.toBe(5);
    firstRender.unmount();
    localStorage.setItem('count', 5);
    const secondRender = render(<RaffleContainer />);
    expect(
      parseInt(secondRender.getByLabelText(/Number of winners/i).value, 10),
    ).toBe(5);
  });

  it('submits and persists data to localStorage (if available)', async () => {
    const { getByLabelText, getByText } = render(<RaffleContainer />);
    mockAxios.get.mockImplementationOnce(() =>
      Promise.resolve({ data: { winners: mockWinners } }),
    );

    expect(mockAxios.get).not.toHaveBeenCalled();
    expect(localStorage.getItem('count')).toBeNull();

    fillOutForm({ getByLabelText });
    await submitForm({ getByText });

    await wait(() => {
      const countInStorage = localStorage.getItem('count');
      expect(countInStorage).toBe(params.count);
      expect(mockAxios.get).toHaveBeenCalledWith(expect.any(String), {
        params,
      });
    });
  });

  it('shows an error message on error', async () => {
    const { getByLabelText, getByText } = render(<RaffleContainer />);
    const errorMessage = 'garbage';
    mockAxios.get.mockImplementationOnce(() => Promise.resolve(errorMessage));

    fillOutForm({ getByLabelText });
    await submitForm({ getByText });

    await waitForElement(() => getByText(/Malformed response/));
  });

  it('resets the form on reset button click', async () => {
    const { getByLabelText, getByText } = render(<RaffleContainer />);
    mockAxios.get.mockImplementationOnce(() =>
      Promise.resolve({ data: { winners: mockWinners } }),
    );

    fillOutForm({ getByLabelText });
    await submitForm({ getByText });

    await waitForElement(() => getByText(mockWinners[0].name));
    fireEvent.click(getByText('Reset'));

    // wait for the reset to resolve, then assert
    await wait(() => {
      expect(() => getByText(mockWinners[0].name)).toThrow();
    });
  });

  it('selects current meetup input text on focus', () => {
    const { getByLabelText } = render(<RaffleContainer />);
    const meetupInput = getByLabelText(/Meetup name/);

    expect(meetupInput.selectionStart).toBe(0);
    expect(meetupInput.selectionEnd).toBe(0);

    fillOutForm({ getByLabelText });
    fireEvent.focus(meetupInput);

    expect(meetupInput.selectionStart).toBe(0);
    expect(meetupInput.selectionEnd).toBe(meetupInput.value.length);
  });
});
