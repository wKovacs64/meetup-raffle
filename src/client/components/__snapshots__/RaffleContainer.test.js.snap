// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaffleContainer renders 1`] = `
<main
  class="flex flex-column flex-grow-1 flex-shrink-0 ph3 w-100 mw6-m mw7-l self-center-ns"
>
  <form
    action="#"
  >
    <div
      class="mt3 mb4 mv4-ns"
    >
      <label
        class="f4 f3-ns lh-copy dark-blue"
        for="meetup"
      >
        Meetup name (from your URL):
      </label>
      <input
        class="input-reset f4 f3-ns near-black ba bw3 b--moon-gray mt3 pa2 w-100"
        id="meetup"
        name="meetup"
        placeholder="required"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-column flex-row-ns items-end-ns justify-around-ns mv4"
    >
      <div
        class="self-center"
      >
        <div
          class="mb3"
        >
          <label
            class="f4 f3-ns lh-copy dark-blue"
            for="count"
          >
            Number of winners:
          </label>
        </div>
        <span
          class="flex flex-row"
        >
          <button
            aria-label="decrement"
            class="bn bg-transparent h3 w3 pointer silver"
            data-testid="decrement-button"
            disabled=""
            type="button"
          >
            <svg
              fill="currentColor"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 40 40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="m22.5 17.5v-5h-5v5h-5l7.5 10 7.5-10h-5z"
                />
              </g>
            </svg>
          </button>
          <input
            class="tc near-black w3 f3 pv1"
            id="count"
            name="count"
            pattern="[0-9]*"
            type="text"
            value="1"
          />
          <button
            aria-label="increment"
            class="bn bg-transparent h3 w3 pointer near-black"
            data-testid="increment-button"
            type="button"
          >
            <svg
              fill="currentColor"
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 40 40"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g>
                <path
                  d="m20 12.5l-7.5 10h5v5h5v-5h5l-7.5-10z"
                />
              </g>
            </svg>
          </button>
        </span>
      </div>
      <div
        class="mv4 mv0-ns"
      >
        <button
          class="db center-ns w-100 w5-ns f5 f4-ns b input-reset ba near-black b--near-black bg-white hover-bg-moon-gray pointer ph5 pv3 shadow-5"
          type="submit"
        >
          Draw
        </button>
      </div>
    </div>
  </form>
</main>
`;
